### PR TITLE
Add -O3 option in build.jl

### LINKED ISSUE #104: Compile HSL with -03 optimization
After some discussions about performances with @geoffroyleconte, I compared OpenBLAS32 / MKL backend and a basic compilation with a " -O3" compilation.
We should add `-03` after `gfortran` and `gcc` by default.
```julia
const HSL_FC = haskey(ENV, "HSL_FC") ? ENV["HSL_FC"] : "gfortran -O3"
const HSL_F77 = haskey(ENV, "HSL_F77") ? ENV["HSL_F77"] : HSL_FC
const HSL_CC = haskey(ENV, "HSL_CC") ? ENV["HSL_CC"] : "gcc -O3"
```

```julia
# Current version of HSL
using HSL, MatrixMarket, SuiteSparseMatrixCollection
using LinearAlgebra, Printf, BenchmarkTools

ssmc = ssmc_db(verbose=false)
matrix = ssmc_matrices(ssmc, "Boeing", "pwtk")
path = fetch_ssmc(matrix, format="MM")

n = matrix.nrows[1]
A = MatrixMarket.mmread(joinpath(path[1], "$(matrix.name[1]).mtx"))
b = ones(n)
b_norm = norm(b)

# Solve Ax = b.
LDL = @btime Ma57($A)           #  7.566 s (36 allocations: 343.44 MiB)
@btime ma57_factorize($LDL)     # 39.155 s (2 allocations: 851.30 KiB)
@btime ma57_solve($LDL, $b)     # 497.909 ms (6 allocations: 4.16 MiB)

import LinearAlgebra, MKL_jll
LinearAlgebra.BLAS.lbt_forward(MKL_jll.libmkl_rt_path, clear=true, verbose=true)

# Solve Ax = b.
LDL = @btime Ma57($A)          #  7.466 s (36 allocations: 343.44 MiB)
@btime ma57_factorize($LDL)    # 25.038 s (2 allocations: 851.30 KiB)
@btime ma57_solve($LDL, $b)    # 230.605 ms (6 allocations: 4.16 MiB)
```

```julia
# HSL compiled with -O3
using HSL, MatrixMarket, SuiteSparseMatrixCollection
using LinearAlgebra, Printf, BenchmarkTools

ssmc = ssmc_db(verbose=false)
matrix = ssmc_matrices(ssmc, "Boeing", "pwtk")
path = fetch_ssmc(matrix, format="MM")

n = matrix.nrows[1]
A = MatrixMarket.mmread(joinpath(path[1], "$(matrix.name[1]).mtx"))
b = ones(n)
b_norm = norm(b)

# Solve Ax = b.
LDL = @btime Ma57($A)           #  3.123 s (36 allocations: 343.44 MiB)
@btime ma57_factorize($LDL)     # 14.857 s (2 allocations: 851.30 KiB)
@btime ma57_solve($LDL, $b)     # 314.188 ms (6 allocations: 4.16 MiB)

import LinearAlgebra, MKL_jll
LinearAlgebra.BLAS.lbt_forward(MKL_jll.libmkl_rt_path, clear=true, verbose=true)

# Solve Ax = b.
LDL = @btime Ma57($A)          # 3.345 s (36 allocations: 343.44 MiB)
@btime ma57_factorize($LDL)    # 9.488 s (2 allocations: 851.30 KiB)
@btime ma57_solve($LDL, $b)    # 186.227 ms (6 allocations: 4.16 MiB)
```

### PR DESCRIPTION
**Pull request by amontoison**
close #104

### CODE REVIEW

**Comment by amontoison** [comment]
@JSOBot runtests

File: deps/build.jl [hunk 1]
```diff
@@ -95,9 +95,10 @@ const hsl_ma97_archive =
 ##############################
 const hsl_archives = [hsl_ma57_archive, hsl_ma97_archive]
 
-const HSL_FC = haskey(ENV, "HSL_FC") ? ENV["HSL_FC"] : "gfortran"
-const HSL_F77 = haskey(ENV, "HSL_F77") ? ENV["HSL_F77"] : HSL_FC
-const HSL_CC = haskey(ENV, "HSL_CC") ? ENV["HSL_CC"] : "gcc"
+const HSL_OPT = " -O3"
+const HSL_FC = (haskey(ENV, "HSL_FC") ? ENV["HSL_FC"] : "gfortran") * HSL_OPT
+const HSL_F77 = (haskey(ENV, "HSL_F77") ? ENV["HSL_F77"] : "gfortran") * HSL_OPT
```
Comment (amontoison) [inline deps/build.jl:100]:
```suggestion
const HSL_F77 = haskey(ENV, "HSL_F77") ? ENV["HSL_F77"] * HSL_OPT : HSL_FC
```

**Comment by amontoison** [comment]
@JSOBot runtests

**Comment by dpo** [comment]
I meant why not put it directly here: https://github.com/JuliaSmoothOptimizers/HSL.jl/blob/main/deps/build_hsl_ma57.jl#L12 ?

**Comment by amontoison** [comment]
> I meant why not put it directly here: https://github.com/JuliaSmoothOptimizers/HSL.jl/blob/main/deps/build_hsl_ma57.jl#L12 ?

Because I didn't know if it was also needed here: https://github.com/JuliaSmoothOptimizers/HSL.jl/blob/main/deps/build_hsl_ma57.jl#L16

**Comment by dpo** [comment]
No, the second call to gfortran only repackages the object files into a shared library.

**Comment by amontoison** [comment]
@JSOBot runtests

File: deps/build.jl [hunk 2]
```diff
@@ -96,7 +96,7 @@ const hsl_ma97_archive =
 const hsl_archives = [hsl_ma57_archive, hsl_ma97_archive]
 
 const HSL_FC = haskey(ENV, "HSL_FC") ? ENV["HSL_FC"] : "gfortran"
-const HSL_F77 = haskey(ENV, "HSL_F77") ? ENV["HSL_F77"] : HSL_FC
+const HSL_F77 = haskey(ENV, "HSL_F77") ? ENV["HSL_F77"] : ENV["HSL_FC"]
```
Comment (amontoison) [inline deps/build.jl:99]:
```suggestion
const HSL_F77 = haskey(ENV, "HSL_F77") ? ENV["HSL_F77"] : HSL_FC
```

**Comment by amontoison** [comment]
@JSOBot runtests

**Comment by dpo** [comment]
Thank you

**Comment by geoffroyleconte** [comment]
I did not see any improvement when using this on frontal22 with MA57. Is there something I should do differently, apart from setting `ENV["HSL_FC"]`, `ENV["HSL_CC"]`, and `ENV["HSL_MA57_PATH"]` ?

**Comment by dpo** [comment]
Silly question, but did you make sure to recompile?

**Comment by geoffroyleconte** [comment]
I added the new version of this repo, set the environment variables and then `]build HSL`.

**Pull request merged**